perf/src/recycler.rs

```diff
@@ -289,63 +88,60 @@ impl<T: Default + Reset + Sized> Recycler<T> {
         new
     }
 
-    pub fn allocate(&self) -> Option<T> {
-        let (allocation_decision, shrink_output) = {
-            let mut object_pool = self
-                .recycler
-                .gc
-                .lock()
-                .expect("recycler lock in pb fn allocate");
-
-            let shrink_output = object_pool.shrink_if_necessary(self.shrink_metric_name);
-
-            // Grab the allocation decision and shrinking stats, do the expensive
-            // allocations/deallocations outside of the lock.
-            (object_pool.make_allocation_decision(), shrink_output)
-        };
-
-        if let Some((mut shrink_stats, shrink_removed_objects)) = shrink_output {
-            let mut shrink_removed_object_elapsed = Measure::start("shrink_removed_object_elapsed");
-            drop(shrink_removed_objects);
-            shrink_removed_object_elapsed.stop();
-            shrink_stats.drop_elapsed = shrink_removed_object_elapsed.as_us();
-            shrink_stats.report(self.shrink_metric_name);
-        }
-
-        match allocation_decision {
-            AllocationDecision::Reuse(mut reused_object) => {
+    pub fn allocate(&self, name: &'static str) -> T {
+        {
+            const RECYCLER_SHRINK_WINDOW_HALF: usize = RECYCLER_SHRINK_WINDOW / 2;
+            const RECYCLER_SHRINK_WINDOW_SUB_ONE: usize = RECYCLER_SHRINK_WINDOW - 1;
+            let mut gc = self.recycler.gc.lock().unwrap();
+            // Update the exponential moving average of gc.len().
+            self.recycler.size_factor.store(
+                self.recycler
+                    .size_factor
+                    .load(Ordering::Acquire)
+                    .saturating_mul(RECYCLER_SHRINK_WINDOW_SUB_ONE)
+                    .saturating_add(RECYCLER_SHRINK_WINDOW_HALF)
+                    .checked_div(RECYCLER_SHRINK_WINDOW)
+                    .unwrap()
+                    .saturating_add(gc.len()),
+                Ordering::Release,
```

Comment:
For the sake of statistics noobs like me 👶 , could we add a definition for the exponential moving average then add some comments on which constants used/ intuition for picking those constants?

From my understanding if the value at iteration `t` is `V_t`, then the exponential average `A` at iteration `t` is computed as for some coefficient `x`:

```
A_t = xV_{t} + (1 - x) * A_{t - 1}
```

From my naive understanding, I don't quite fully grasp the current implementation: https://github.com/solana-labs/solana/pull/16457/files#diff-798271cb43dca47253c06d8ab6888d8bf4fb385fc8e74cc2cd70ac1b3b664ebfR97-R106

It seems to me that in the above code if `gc.len()` remains constant then the `size_factor` blows up in size quickly because we're adding `.saturating_add(gc.len())` every time.

 For instance if `gc.len() == RECYCLER_SHRINK_WINDOW` remains constant then this current calculation of `size_factor` would increase by many multiples every iteration.

Reply:
Ah I see, thanks for the explanation, that was super helpful for my understanding.

Could we add this explanation as a comment in the code?

Reply:
added comments